src/NServiceBus.Core/Routing/ApplyReplyToAddressBehavior.cs

```diff
@@ -30,17 +31,26 @@ public Task Invoke(IOutgoingLogicalMessageContext context, Func<IOutgoingLogical
             {
                 throw new InvalidOperationException("Cannot route a reply to a specific instance because an endpoint instance discriminator was not configured for the destination endpoint. It can be specified via EndpointConfiguration.MakeInstanceUniquelyAddressable(string discriminator).");
             }
-            context.Headers[Headers.ReplyToAddress] = ApplyUserOverride(publicReturnAddress ?? sharedQueue, state);
 
-            //Legacy distributor logic
-            IncomingMessage incomingMessage;
-            if (context.TryGetIncomingPhysicalMessage(out incomingMessage) && incomingMessage.Headers.ContainsKey(LegacyDistributorHeaders.WorkerSessionId))
+            var configuredReturnAddress = publicReturnAddress ?? sharedQueue;
+            if (distributorAddress != null && UseDistributorAddress(context))
```

Comment:
I had to change the implementation slightly from the propsed fix by @SzymonPobiega , please have a close look.